src/xma/include/lib/xma_buffer.hpp

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright (C) 2019 Xilinx, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#ifndef _XMA_BUFFERCLS_H_
+#define _XMA_BUFFERCLS_H_
+
+#include <cstdint>
+#include <cstring>
+#include <string>
+#include <atomic>
+#include "xrt.h"
+#include "core/include/experimental/xrt_bo.h"
+
+namespace xma_core {
+namespace plg {
+
+using memory_group = xrtMemoryGroup;
+
+/**
+ * class buffer : abstraction for xma buffer
+ */
+class buffer
+{
+private:
+    void*    dummy{nullptr};
+    uint8_t* data{nullptr};
+    xrt::bo  xrt_bo;
+    int32_t  bank_index;
+    int32_t  dev_index;
+    std::atomic<int32_t> ref_cnt{0};//For use by plugins when shared by plugins
+    bool     device_only_buffer;
+    uint32_t reserved[4];
+
+public:
+  int32_t
+  read_ddr(int32_t offset, uint64_t size) const;
+
+  int32_t
+  write_ddr(int32_t offset, uint64_t size) const;
+
+  buffer(xclDeviceHandle dhdl, xrt::bo::flags flags, memory_group grp, int32_t dev_idx, int32_t bank_idx, uint64_t sz, bool dev_only);
```

Comment:
Why dev_only?  Isn't that part of flags? 
And what is bank_idx compared to grp?
And what is the purpose of dev_idx?

Please comment these APIs so then intent is clear.

Reply:
Nothing wrong in having it here too. No rule says info can not be shared between different objects. It is good to have it here too.

Reply:
Please explain why duplicate information is good to carry around?  I understand caching of accessed data, but I don't understand an API that takes the same information to multiple ways.   

Here I suspect that `flags` already reflects `dev_only`, and that `bank_idx` is identical to` grp` and that `dev_idx` is not needed.  

Of course I have no way of knowing for sure, because there is no documentation as to the meaning of the constructor arguments.

Reply:
yes it is caching of accessed data. changed the constructor api